File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        if (head == NULL) return false;
        ListNode *p1 = head, *p2 = head->next;
        while (p1 != NULL && p2 != NULL) {
            if (p1 != NULL) p1 = p1->next;
            if (p2 != NULL) p2 = p2->next;
            if (p2 != NULL) p2 = p2->next; // move p2 twice
            if (p1 == p2) return true;
        }
        return false;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        sz = 0
        temp = head
        while temp:
            sz += 1
            temp = temp.next
        
        remove = sz - n
        if remove == 0: return head.next

        cur = head
        for i in range(sz - 1):
            if (i + 1) == remove:
                cur.next = cur.next.next
                break
            cur = cur.next
        return head
        